### PR TITLE
Personalized redirect

### PR DESCRIPTION
**Pull request by torotil**
This includes:

* A vue app to deal with the configuration.
* A configuration storage API in campaignion_action.
* The integration in the wizard ThankyouStep.
* The actual implementation of the redirect behaviour in campaignion_action.

### CODE REVIEW

File: campaignion_action/src/ActionBase.php [hunk 1]
```diff
@@ -121,11 +121,11 @@ public function redirect(Submission $submission, $delta) {
       case 'redirect':
         $redirects = Redirect::byNid($this->node->nid, $delta);
         foreach ($redirects as $r) {
-        if ($r->checkFilters($submission)) {
-          return $r->normalized();
+          if ($r->checkFilters($submission)) {
+            return $r->normalized();
+          }
         }
         break;
```
Comment (alex-mo) [inline campaignion_action/src/ActionBase.php]: dangling `break`. can be dropped

**Comment by alex-mo** [comment]
* +1 for the usage of constants
* +1 for meaningful comments
* `break` unnecessary? (see inline comment)
* probably also test for the non-check optin return `FALSE` as this should not happen
* is there a specific reasoning making the functions `static`? (the commit with the `#process` fn)

**Review by alex-mo — CHANGES_REQUESTED** [review]

File: campaignion_action/src/ActionBase.php (hunk 1 above)
Comment (torotil) [reply campaignion_action/src/ActionBase.php]: I’d vote to keep. Think of someone editing the code in the future in a way that doesn’t lead to a return in any case …

**Comment by torotil** [comment]
Yes, there is a reason to use static functions here: The `#process` property is stored in the form cache (the whole `$form` is cached there). So adding a `[$this, 'addRedirectSettigs']` as callback would put the whole form object there. This means:

1. The step object needs to be serialisable.
2. We add lots of redundant data to the form cache.
3. We have another (older) copy of the step object in there in later requests. Sometimes the "new" one is used, sometimes the old one is used. This adds loads of complexity.

**Comment by torotil** [comment]
Added the additional test for the negated opt-in filter.

File: campaignion_action/src/ActionBase.php (hunk 1 above)
Comment (alex-mo) [reply campaignion_action/src/ActionBase.php]: ok with me...

**Review by alex-mo — APPROVED** [review]

**Pull request merged**